cli/azd/pkg/project/service_target_ml_endpoint.go

```diff
@@ -0,0 +1,282 @@
+			modelVersion, err := m.aiHelper.CreateModelVersion(ctx, workspaceScope, serviceConfig, endpointConfig.Model)
+			if err != nil {
+				task.SetError(err)
+				return
+			}
+
+			deployResult.Model = modelVersion
+		}
+
+		// Deploy to endpoint
+		if endpointConfig.Deployment != nil {
+			task.SetProgress(NewServiceProgress("Deploying to endpoint"))
+			endpointName := filepath.Base(targetResource.ResourceName())
+			onlineDeployment, err := m.aiHelper.DeployToEndpoint(
+				ctx,
+				workspaceScope,
+				serviceConfig,
+				endpointName,
+				endpointConfig,
+			)
+			if err != nil {
+				task.SetError(err)
+				return
+			}
+
+			task.SetProgress(NewServiceProgress("Updating traffic"))
+			_, err = m.aiHelper.UpdateTraffic(ctx, workspaceScope, endpointName, *onlineDeployment.Name)
+			if err != nil {
+				task.SetError(fmt.Errorf("failed updating traffic: %w", err))
+				return
+			}
+
+			task.SetProgress(NewServiceProgress("Removing old deployments"))
+			if err := m.aiHelper.DeletePreviousDeployments(ctx, workspaceScope, endpointName); err != nil {
+				task.SetError(fmt.Errorf("failed deleting previous deployments: %w", err))
+				return
+			}
+
+			deployResult.Deployment = onlineDeployment
+		}
+
+		endpoints, err := m.Endpoints(ctx, serviceConfig, targetResource)
+		if err != nil {
+			task.SetError(err)
+			return
+		}
+
+		if err := m.envManager.Save(ctx, m.env); err != nil {
+			task.SetError(fmt.Errorf("failed saving environment: %w", err))
+			return
+		}
+
+		task.SetResult(&ServiceDeployResult{
+			Details:   deployResult,
+			Package:   servicePackage,
+			Endpoints: endpoints,
+		})
+	})
+}
+
+// Endpoints returns the endpoints for the service
+func (m *machineLearningEndpointTarget) Endpoints(
+	ctx context.Context,
+	serviceConfig *ServiceConfig,
+	targetResource *environment.TargetResource,
+) ([]string, error) {
+	workspaceScope, err := m.getWorkspaceScope(serviceConfig, targetResource)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointName := filepath.Base(targetResource.ResourceName())
+	onlineEndpoint, err := m.aiHelper.GetEndpoint(ctx, workspaceScope, endpointName)
+	if err != nil {
+		return nil, err
+	}
+
+	return []string{
+		fmt.Sprintf("Scoring: %s", *onlineEndpoint.Properties.ScoringURI),
+		fmt.Sprintf("Swagger: %s", *onlineEndpoint.Properties.SwaggerURI),
+	}, nil
+}
+
+// getWorkspaceScope returns the scope for the workspace
+func (m *machineLearningEndpointTarget) getWorkspaceScope(
+	serviceConfig *ServiceConfig,
+	targetResource *environment.TargetResource,
+) (*ai.Scope, error) {
+	endpointConfig, err := ai.ParseConfig[ai.EndpointDeploymentConfig](serviceConfig.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	workspaceName, err := endpointConfig.Workspace.Envsubst(m.env.Getenv)
+	if err != nil {
+		return nil, err
+	}
+
+	// Workspace name can come from the following:
+	// 1. The workspace field in the endpoint service config
+	// 2. The AZUREML_AI_PROJECT_NAME environment variable
```

Comment:
Do we also want to support something like `SERVICE_<service-name>_AI_PROJECT_NAME` so different services can be tied to different AI projects?  We've used that pattern elsewhere.

Reply:
At this point I haven't heard of a requirement to deploy services out to different AI projects but I'm sure someone may try it.  At that point I would also consider adding a `project` configuration param to the service host config similar to the `workspace` that we have today.

We could then easily support the env vars model as defaults if we wanted to go down that route.

Reply:
Workspace ~= project.  In AI studio they call it a `project` but the underlying Azure resource is still an ML Workspace with some specialized `kind`.

Since the `workspace` param in the `azure.yaml` already supports environment variable substitution they could support multiple projects via different environment variables instead of `azd` having to come up with a specialized naming convention.  Or they can just hard code a string if they really wanted.